File: alice/megamind/example_apps/java/src/main/java/ru/yandex/alice/megamind/cowsay/CustomProtobufHttpMessageConverter.java
package ru.yandex.alice.megamind.cowsay;

import com.google.protobuf.CodedOutputStream;
import com.google.protobuf.ExtensionRegistry;
import com.google.protobuf.Message;
import com.google.protobuf.TextFormat;
import com.google.protobuf.util.JsonFormat;
import org.springframework.http.HttpInputMessage;
import org.springframework.http.HttpOutputMessage;
import org.springframework.http.MediaType;
import org.springframework.http.converter.AbstractHttpMessageConverter;
import org.springframework.http.converter.HttpMessageConversionException;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.http.converter.HttpMessageNotWritableException;
import org.springframework.lang.Nullable;
import org.springframework.util.Assert;
import org.springframework.util.ConcurrentReferenceHashMap;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.lang.reflect.Method;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Map;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.http.MediaType.TEXT_PLAIN;

public class CustomProtobufHttpMessageConverter extends AbstractHttpMessageConverter<Message> {

    /**
     * The default charset used by the converter.
     */
    public static final Charset DEFAULT_CHARSET = StandardCharsets.UTF_8;

    /**
     * The media-type for protobuf {@code application/x-protobuf}.
     */
    public static final MediaType PROTOBUF = new MediaType("application", "protobuf");
    public static final MediaType X_PROTOBUF = new MediaType("application", "x-protobuf");

    /**
     * The HTTP header containing the protobuf schema.
     */
    public static final String X_PROTOBUF_SCHEMA_HEADER = "X-Protobuf-Schema";

    /**
     * The HTTP header containing the protobuf message.
     */
    public static final String X_PROTOBUF_MESSAGE_HEADER = "X-Protobuf-Message";


    private static final Map<Class<?>, Method> methodCache = new ConcurrentReferenceHashMap<>();

    private final ExtensionRegistry extensionRegistry;

    private final ProtobufFormatSupport protobufFormatSupport;

    public CustomProtobufHttpMessageConverter(
            @Nullable JsonFormat.Parser parser,
            @Nullable JsonFormat.Printer printer
    ) {
        this(parser, printer, null);
    }

    public CustomProtobufHttpMessageConverter(
            @Nullable JsonFormat.Parser parser,
            @Nullable JsonFormat.Printer printer,
            @Nullable ExtensionRegistry extensionRegistry
    ) {

        this.protobufFormatSupport = new ProtobufJavaUtilSupport((parser != null ? parser : JsonFormat.parser()), (printer != null ? printer : JsonFormat.printer()));

        setSupportedMediaTypes(Arrays.asList(this.protobufFormatSupport.supportedMediaTypes()));

        this.extensionRegistry = (extensionRegistry == null ? ExtensionRegistry.newInstance() : extensionRegistry);
    }


    @Override
    protected boolean supports(Class<?> clazz) {
        return Message.class.isAssignableFrom(clazz);
    }

    @Override
    protected MediaType getDefaultContentType(Message message) {
        return PROTOBUF;
    }

    @Override
    protected Message readInternal(Class<? extends Message> clazz, HttpInputMessage inputMessage)
            throws IOException, HttpMessageNotReadableException {

        MediaType contentType = inputMessage.getHeaders().getContentType();
        if (contentType == null) {
            contentType = PROTOBUF;
        }
        Charset charset = contentType.getCharset();
        if (charset == null) {
            charset = DEFAULT_CHARSET;
        }

        Message.Builder builder = getMessageBuilder(clazz);
        if (PROTOBUF.isCompatibleWith(contentType) || X_PROTOBUF.isCompatibleWith(contentType)) {
            builder.mergeFrom(inputMessage.getBody(), this.extensionRegistry);
        } else if (TEXT_PLAIN.isCompatibleWith(contentType)) {
            InputStreamReader reader = new InputStreamReader(inputMessage.getBody(), charset);
            TextFormat.merge(reader, this.extensionRegistry, builder);
        } else if (this.protobufFormatSupport != null) {
            this.protobufFormatSupport.merge(
                    inputMessage.getBody(), charset, contentType, this.extensionRegistry, builder);
        }
        return builder.build();
    }

    /**
     * Create a new {@code Message.Builder} instance for the given class.
     * <p>This method uses a ConcurrentReferenceHashMap for caching method lookups.
     */
    private Message.Builder getMessageBuilder(Class<? extends Message> clazz) {
        try {
            Method method = methodCache.get(clazz);
            if (method == null) {
                method = clazz.getMethod("newBuilder");
                methodCache.put(clazz, method);
            }
            return (Message.Builder) method.invoke(clazz);
        } catch (Exception ex) {
            throw new HttpMessageConversionException(
                    "Invalid Protobuf Message type: no invocable newBuilder() method on " + clazz, ex);
        }
    }


    @Override
    protected boolean canWrite(@Nullable MediaType mediaType) {
        return (super.canWrite(mediaType) ||
                (this.protobufFormatSupport != null && this.protobufFormatSupport.supportsWriteOnly(mediaType)));
    }

    @Override
    protected void writeInternal(Message message, HttpOutputMessage outputMessage)
            throws IOException, HttpMessageNotWritableException {

        MediaType contentType = outputMessage.getHeaders().getContentType();
        if (contentType == null) {
            contentType = getDefaultContentType(message);
            Assert.state(contentType != null, "No content type");
        }
        Charset charset = contentType.getCharset();
        if (charset == null) {
            charset = DEFAULT_CHARSET;
        }

        if (PROTOBUF.isCompatibleWith(contentType) || X_PROTOBUF.isCompatibleWith(contentType)) {
            setProtoHeader(outputMessage, message);
            CodedOutputStream codedOutputStream = CodedOutputStream.newInstance(outputMessage.getBody());
            message.writeTo(codedOutputStream);
            codedOutputStream.flush();
        } else if (TEXT_PLAIN.isCompatibleWith(contentType)) {
            OutputStreamWriter outputStreamWriter = new OutputStreamWriter(outputMessage.getBody(), charset);
            TextFormat.print(message, outputStreamWriter);
            outputStreamWriter.flush();
            outputMessage.getBody().flush();
        } else if (this.protobufFormatSupport != null) {
            this.protobufFormatSupport.print(message, outputMessage.getBody(), contentType, charset);
            outputMessage.getBody().flush();
        }
    }

    /**
     * Set the "X-Protobuf-*" HTTP headers when responding with a message of
     * content type "application/x-protobuf"
     * <p><b>Note:</b> <code>outputMessage.getBody()</code> should not have been called
     * before because it writes HTTP headers (making them read only).</p>
     */
    private void setProtoHeader(HttpOutputMessage response, Message message) {
        response.getHeaders().set(X_PROTOBUF_SCHEMA_HEADER, message.getDescriptorForType().getFile().getName());
        response.getHeaders().set(X_PROTOBUF_MESSAGE_HEADER, message.getDescriptorForType().getFullName());
    }


    /**
     * Protobuf format support.
     */
    interface ProtobufFormatSupport {

        MediaType[] supportedMediaTypes();

        boolean supportsWriteOnly(@Nullable MediaType mediaType);

        void merge(InputStream input, Charset charset, MediaType contentType,
                   ExtensionRegistry extensionRegistry, Message.Builder builder)
                throws IOException, HttpMessageConversionException;

        void print(Message message, OutputStream output, MediaType contentType, Charset charset)
                throws IOException, HttpMessageConversionException;
    }


    /**
     * {@link ProtobufFormatSupport} implementation used when
     * {@code com.google.protobuf.util.JsonFormat} is available.
     */
    private static class ProtobufJavaUtilSupport implements ProtobufFormatSupport {

        private final JsonFormat.Parser parser;

        private final JsonFormat.Printer printer;

        public ProtobufJavaUtilSupport(@Nullable JsonFormat.Parser parser, @Nullable JsonFormat.Printer printer) {
            this.parser = (parser != null ? parser : JsonFormat.parser());
            this.printer = (printer != null ? printer : JsonFormat.printer());
        }

        @Override
        public MediaType[] supportedMediaTypes() {
            return new MediaType[] {PROTOBUF, X_PROTOBUF, TEXT_PLAIN, APPLICATION_JSON};
        }

        @Override
        public boolean supportsWriteOnly(@Nullable MediaType mediaType) {
            return false;
        }

        @Override
        public void merge(InputStream input, Charset charset, MediaType contentType,
                          ExtensionRegistry extensionRegistry, Message.Builder builder)
                throws IOException, HttpMessageConversionException {

            if (contentType.isCompatibleWith(APPLICATION_JSON)) {
                InputStreamReader reader = new InputStreamReader(input, charset);
                this.parser.merge(reader, builder);
            } else {
                throw new HttpMessageConversionException(
                        "protobuf-java-util does not support parsing " + contentType);
            }
        }

        @Override
        public void print(Message message, OutputStream output, MediaType contentType, Charset charset)
                throws IOException, HttpMessageConversionException {

            if (contentType.isCompatibleWith(APPLICATION_JSON)) {
                OutputStreamWriter writer = new OutputStreamWriter(output, charset);
                this.printer.appendTo(message, writer);
                writer.flush();
            } else {
                throw new HttpMessageConversionException(
                        "protobuf-java-util does not support printing " + contentType);
            }
        }
    }
}

File: alice/memento/src/main/kotlin/controller/CustomProtobufHttpMessageConverter.kt
package ru.yandex.alice.memento.controller

import com.google.protobuf.CodedOutputStream
import com.google.protobuf.ExtensionRegistry
import com.google.protobuf.Message
import com.google.protobuf.TextFormat
import com.google.protobuf.util.JsonFormat
import org.apache.logging.log4j.LogManager
import org.springframework.http.HttpInputMessage
import org.springframework.http.HttpOutputMessage
import org.springframework.http.MediaType
import org.springframework.http.converter.AbstractHttpMessageConverter
import org.springframework.http.converter.HttpMessageConversionException
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.http.converter.HttpMessageNotWritableException
import org.springframework.util.ConcurrentReferenceHashMap
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.OutputStream
import java.io.OutputStreamWriter
import java.lang.reflect.Method
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

internal class CustomProtobufHttpMessageConverter
@JvmOverloads constructor(
    parser: JsonFormat.Parser?,
    printer: JsonFormat.Printer?,
    extensionRegistry: ExtensionRegistry? = null
) : AbstractHttpMessageConverter<Message>() {

    private val extensionRegistry = extensionRegistry ?: ExtensionRegistry.newInstance()
    private val protobufFormatSupport = ProtobufJavaUtilSupport(
        parser ?: JsonFormat.parser(),
        printer ?: JsonFormat.printer()
    )

    init {
        supportedMediaTypes = listOf(*protobufFormatSupport.supportedMediaTypes())
    }

    override fun supports(clazz: Class<*>): Boolean {
        return Message::class.java.isAssignableFrom(clazz)
    }

    override fun getDefaultContentType(message: Message): MediaType {
        return PROTOBUF
    }

    @Throws(IOException::class, HttpMessageNotReadableException::class)
    override fun readInternal(clazz: Class<out Message>, inputMessage: HttpInputMessage): Message {
        val builder: Message.Builder
        try {
            val contentType = inputMessage.headers.contentType ?: PROTOBUF
            val charset = contentType.charset ?: DEFAULT_CHARSET

            builder = getMessageBuilder(clazz)
            if (PROTOBUF.isCompatibleWith(contentType) || X_PROTOBUF.isCompatibleWith(contentType)) {
                builder.mergeFrom(inputMessage.body, extensionRegistry)
            } else if (MediaType.TEXT_PLAIN.isCompatibleWith(contentType)) {
                val reader = InputStreamReader(inputMessage.body, charset)
                TextFormat.merge(reader, extensionRegistry, builder)
            } else if (protobufFormatSupport != null) {
                protobufFormatSupport.merge(
                    inputMessage.body, charset, contentType, extensionRegistry, builder
                )
            }
        } catch (e: Exception) {
            Companion.logger.error("Read protobuf message exception", e)
            throw e
        }
        return builder.build()
    }

    /**
     * Create a new `Message.Builder` instance for the given class.
     *
     * This method uses a ConcurrentReferenceHashMap for caching method lookups.
     */
    private fun getMessageBuilder(clazz: Class<out Message>): Message.Builder {
        return try {
            var method = METHOD_CACHE[clazz]
            if (method == null) {
                method = clazz.getMethod("newBuilder")
                METHOD_CACHE[clazz] = method
            }
            method!!.invoke(clazz) as Message.Builder
        } catch (ex: Exception) {
            throw HttpMessageConversionException(
                "Invalid Protobuf Message type: no invocable newBuilder() method on $clazz", ex
            )
        }
    }

    override fun canWrite(mediaType: MediaType?): Boolean {
        return super.canWrite(mediaType) ||
            protobufFormatSupport != null && protobufFormatSupport.supportsWriteOnly(mediaType)
    }

    @Throws(IOException::class, HttpMessageNotWritableException::class)
    override fun writeInternal(message: Message, outputMessage: HttpOutputMessage) {
        val contentType = outputMessage.headers.contentType ?: getDefaultContentType(message)
        val charset = contentType.charset ?: DEFAULT_CHARSET

        if (PROTOBUF.isCompatibleWith(contentType) || X_PROTOBUF.isCompatibleWith(contentType)) {
            setProtoHeader(outputMessage, message)
            val codedOutputStream = CodedOutputStream.newInstance(outputMessage.body)
            message.writeTo(codedOutputStream)
            codedOutputStream.flush()
        } else if (MediaType.TEXT_PLAIN.isCompatibleWith(contentType)) {
            val outputStreamWriter = OutputStreamWriter(outputMessage.body, charset)
            TextFormat.printer().print(message, outputStreamWriter)
            outputStreamWriter.flush()
            outputMessage.body.flush()
        } else if (protobufFormatSupport != null) {
            protobufFormatSupport.print(message, outputMessage.body, contentType, charset)
            outputMessage.body.flush()
        }
    }

    /**
     * Set the "X-Protobuf-*" HTTP headers when responding with a message of
     * content type "application/x-protobuf"
     *
     * **Note:** `outputMessage.getBody()` should not have been called
     * before because it writes HTTP headers (making them read only).
     */
    private fun setProtoHeader(response: HttpOutputMessage, message: Message) {
        response.headers[X_PROTOBUF_SCHEMA_HEADER] = message.descriptorForType.file.name
        response.headers[X_PROTOBUF_MESSAGE_HEADER] = message.descriptorForType.fullName
    }

    /**
     * Protobuf format support.
     */
    internal interface ProtobufFormatSupport {
        fun supportedMediaTypes(): Array<MediaType>
        fun supportsWriteOnly(mediaType: MediaType?): Boolean

        @Throws(IOException::class, HttpMessageConversionException::class)
        fun merge(
            input: InputStream,
            charset: Charset,
            contentType: MediaType,
            extensionRegistry: ExtensionRegistry,
            builder: Message.Builder
        )

        @Throws(IOException::class, HttpMessageConversionException::class)
        fun print(message: Message, output: OutputStream, contentType: MediaType, charset: Charset)
    }

    /**
     * [ProtobufFormatSupport] implementation used when
     * `com.google.protobuf.util.JsonFormat` is available.
     */
    private class ProtobufJavaUtilSupport(parser: JsonFormat.Parser?, printer: JsonFormat.Printer?) : ProtobufFormatSupport {
        private val parser: JsonFormat.Parser = parser ?: JsonFormat.parser()
        private val printer: JsonFormat.Printer = printer ?: JsonFormat.printer()
        override fun supportedMediaTypes(): Array<MediaType> {
            return arrayOf(PROTOBUF, X_PROTOBUF, MediaType.TEXT_PLAIN, MediaType.APPLICATION_JSON)
        }

        override fun supportsWriteOnly(mediaType: MediaType?): Boolean {
            return false
        }

        @Throws(IOException::class, HttpMessageConversionException::class)
        override fun merge(
            input: InputStream,
            charset: Charset,
            contentType: MediaType,
            extensionRegistry: ExtensionRegistry,
            builder: Message.Builder
        ) {
            if (contentType.isCompatibleWith(MediaType.APPLICATION_JSON)) {
                val reader = InputStreamReader(input, charset)
                parser.merge(reader, builder)
            } else {
                throw HttpMessageConversionException(
                    "protobuf-java-util does not support parsing $contentType"
                )
            }
        }

        @Throws(IOException::class, HttpMessageConversionException::class)
        override fun print(message: Message, output: OutputStream, contentType: MediaType, charset: Charset) {
            if (contentType.isCompatibleWith(MediaType.APPLICATION_JSON)) {
                val writer = OutputStreamWriter(output, charset)
                printer.appendTo(message, writer)
                writer.flush()
            } else {
                throw HttpMessageConversionException(
                    "protobuf-java-util does not support printing $contentType"
                )
            }
        }
    }

    companion object {
        private val logger = LogManager.getLogger()

        /**
         * The default charset used by the converter.
         */
        val DEFAULT_CHARSET = StandardCharsets.UTF_8

        /**
         * The media-type for protobuf `application/x-protobuf`.
         */
        val PROTOBUF = MediaType("application", "protobuf")
        val X_PROTOBUF = MediaType("application", "x-protobuf")

        /**
         * The HTTP header containing the protobuf schema.
         */
        const val X_PROTOBUF_SCHEMA_HEADER = "X-Protobuf-Schema"

        /**
         * The HTTP header containing the protobuf message.
         */
        const val X_PROTOBUF_MESSAGE_HEADER = "X-Protobuf-Message"
        private val METHOD_CACHE: MutableMap<Class<*>, Method> = ConcurrentReferenceHashMap()
    }
}
